src/lib.rs

```diff
@@ -208,13 +208,60 @@ where
   _p_c2: PhantomData<C2>,
 }
 
-impl<G1, G2, C1, C2> SimpleDigestible for PublicParams<G1, G2, C1, C2>
+impl<G1, G2, C1, C2> Digestible for PublicParams<G1, G2, C1, C2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
   C1: StepCircuit<G1::Scalar>,
   C2: StepCircuit<G2::Scalar>,
+  G1::Scalar: Serialize,
 {
+  fn write_digestable_bytes<W: Sized + io::Write>(
+    &self,
+    byte_sink: &mut W,
+  ) -> Result<(), io::Error> {
+    self.digest.write_digestable_bytes(byte_sink)
+  }
+
+  fn to_bytes(&self) -> Result<Vec<u8>, io::Error> {
+    let mut byte_sink = vec![];
+    self
+      .F_arity_primary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .F_arity_secondary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .ro_consts_primary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .ro_consts_circuit_primary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self.ck_primary.write_digestable_bytes(&mut byte_sink)?;
+    self
+      .r1cs_shape_primary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .ro_consts_secondary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .ro_consts_circuit_secondary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self.ck_secondary.write_digestable_bytes(&mut byte_sink)?;
+    self
+      .r1cs_shape_secondary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .augmented_circuit_params_primary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self
+      .augmented_circuit_params_secondary
+      .write_digestable_bytes(&mut byte_sink)?;
+    self.digest.write_digestable_bytes(&mut byte_sink)?;
+    self._p_c1.write_digestable_bytes(&mut byte_sink)?;
+    self._p_c2.write_digestable_bytes(&mut byte_sink)?;
```

Comment:
Another important point: do you think this actually binds the public parameters to the shape of the circuit?
Because phantom types beg to differ:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ffd3e4adc7538effc0a088b1b0474734

Reply:
I think these can be left out, exactly because they are *not* actually 'present'. The shapes are there independently, if I'm not mistaken — as straightforward data: https://github.com/lurk-lab/arecibo/blob/b4f3ce80980e6a9f985b5fe9ec44f37b1e21d267/src/lib.rs#L198-L202

Reply:
Yes, that's what I was getting at.